File: lib/tools/quota-tracker.ts
import { adminSupabase } from '@/lib/supabase/admin'

const DEFAULT_PROVIDER = 'api-football'
const DAILY_LIMIT = 100

function todayUTC(): string {
  return new Date().toISOString().split('T')[0]
}

/**
 * Retourne le nombre de requêtes utilisées aujourd'hui pour un provider donné.
 */
export async function getQuotaUsed(provider = DEFAULT_PROVIDER): Promise<number> {
  const { data, error } = await adminSupabase
    .from('api_quota')
    .select('calls_used')
    .eq('date', todayUTC())
    .eq('provider', provider)
    .maybeSingle()

  if (error) throw new Error(`quota-tracker getQuotaUsed: ${error.message}`)
  return data?.calls_used ?? 0
}

/**
 * Incrémente atomiquement le quota du jour via une RPC Supabase.
 * Retourne le nouveau total après incrémentation.
 *
 * La fonction SQL attendue dans Supabase :
 *   CREATE OR REPLACE FUNCTION increment_api_quota(p_date DATE, p_provider TEXT, p_n INT)
 *   RETURNS INT LANGUAGE plpgsql AS $$
 *   DECLARE v_calls INT;
 *   BEGIN
 *     INSERT INTO api_quota (date, provider, calls_used)
 *       VALUES (p_date, p_provider, p_n)
 *     ON CONFLICT (date, provider)
 *       DO UPDATE SET calls_used = api_quota.calls_used + EXCLUDED.calls_used
 *     RETURNING calls_used INTO v_calls;
 *     RETURN v_calls;
 *   END; $$;
 */
export async function incrementQuota(provider = DEFAULT_PROVIDER, n = 1): Promise<number> {
  const { data, error } = await adminSupabase.rpc('increment_api_quota', {
    p_date: todayUTC(),
    p_provider: provider,
    p_n: n,
  })

  if (error) throw new Error(`quota-tracker incrementQuota: ${error.message}`)
  return data as number
}

/**
 * Retourne le nombre de requêtes restantes pour aujourd'hui.
 */
export async function getRemainingQuota(provider = DEFAULT_PROVIDER): Promise<number> {
  const used = await getQuotaUsed(provider)
  return Math.max(0, DAILY_LIMIT - used)
}

File: backend/tools/quota_tracker.py
"""Port de lib/tools/quota-tracker.ts.

Réutilise la RPC Supabase increment_api_quota déjà en place — aucune
migration SQL nécessaire.
"""
from datetime import datetime, timezone

from supabase_client import admin_supabase

DEFAULT_PROVIDER = "api-football"
DAILY_LIMIT = 100


def _today_utc() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def get_quota_used(provider: str = DEFAULT_PROVIDER) -> int:
    res = (
        admin_supabase.table("api_quota")
        .select("calls_used")
        .eq("date", _today_utc())
        .eq("provider", provider)
        .maybe_single()
        .execute()
    )
    return (res.data or {}).get("calls_used", 0)


def increment_quota(provider: str = DEFAULT_PROVIDER, n: int = 1) -> int:
    res = admin_supabase.rpc(
        "increment_api_quota", {"p_date": _today_utc(), "p_provider": provider, "p_n": n}
    ).execute()
    return res.data


def get_remaining_quota(provider: str = DEFAULT_PROVIDER) -> int:
    used = get_quota_used(provider)
    return max(0, DAILY_LIMIT - used)
